### PR TITLE
Providers

### PR DESCRIPTION
**Pull request by dougwig**

### CODE REVIEW

File: a10_neutron_lbaas/handler_base.py [hunk 1]
```diff
@@ -13,15 +13,25 @@
...
     def __init__(self, a10_driver):
         self.a10_driver = a10_driver
-        self.hooks = a10_driver.hooks
         self.openstack_driver = self.a10_driver.openstack_driver
 
+        # XXX(dougwig): is this too racy?
+        if self.a10_driver.provider is None:
+            for prov, driver in self.openstack_driver.plugin.drivers.items():
+                if driver is self.openstack_driver:
+                    self.a10_driver._late_init(prov)
```
Comment (Cedev) [inline a10_neutron_lbaas/handler_base.py]: Why isn't this a10_openstack_lb's job to set it's provider if it's None? Something like

```
def __init__(..., provider=None):
    self._provider = provider

@property
def provider(self):
    """neutron lbaas provider name. The provider name is not available during __init__"""
    if self._provider is None:
        ...
            self._provider = ...
    return self._provider
```

This also gets ugly when we also need to initialize hooks based on config with also depends on the provider.

Comment (dougwig) [inline a10_neutron_lbaas/handler_base.py]: I think provider, config, and hooks all move to handler_base, actually, and out of global init entirely.

**Comment by dougwig** [comment]
check

**Comment by Cedev** [comment]
Looks good so far. I'll look at this again when I have a sharper mind.

File: a10_neutron_lbaas/a10_config.py [hunk 2]
```diff
@@ -77,6 +88,16 @@ def _find_config_dir(self, config_dir):
...
+    # def _load_provider_config(self, provider=None):
+    #     self._outer_config = self._config
+
+    #     if (provider is not None and hasattr(self._config, 'providers') and
+    #             provider in self._config['providers']):
+    #         self._config = self._config.providers[provider]
+
+    #     self._load_config()
+
+    # TODO(dougwig) -- implement fallback _config semantics below
```
Comment (dougwig) [inline a10_neutron_lbaas/a10_config.py]: The above should be deleted.

File: a10_neutron_lbaas/handler_base.py [hunk 3]
```diff
@@ -13,15 +13,32 @@
...
-        self.hooks = a10_driver.hooks
         self.openstack_driver = self.a10_driver.openstack_driver
 
+        if self.a10_driver.provider is None:
+            LATE_INIT_LOCK.acquire()
+            try:
+                if self.a10_driver.provider is None:
+                    for prov, driver in self.openstack_driver.plugin.drivers.items():
+                        if driver is self.openstack_driver:
+                            self.a10_driver._late_init(prov)
```
Comment (dougwig) [inline a10_neutron_lbaas/handler_base.py]: There should be a 'break' after this line.

File: a10_neutron_lbaas/a10_config.py [hunk 4]
```diff
@@ -28,18 +28,29 @@
...
+            return None
+
+        for k, v in d['providers'][provider].items():
+            if isinstance(v, dict):
+                if k not in self.__dict__:
+                    self.__dict__[k] = {}
+                self.__dict__[k].update(v)
+            else:
+                self.__dict__[k] = v
+
```
Comment (Cedev) [inline a10_neutron_lbaas/a10_config.py]: The provider lookup being here means that a `config` passed to `A10Config(config=config, provider='provider1')` won't have provider lookup.

File: a10_neutron_lbaas/handler_base.py [hunk 5]
```diff
@@ -13,15 +13,32 @@
...
 
 class HandlerBase(object):
 
     def __init__(self, a10_driver):
         self.a10_driver = a10_driver
-        self.hooks = a10_driver.hooks
         self.openstack_driver = self.a10_driver.openstack_driver
 
+        if self.a10_driver.provider is None:
+            LATE_INIT_LOCK.acquire()
```
Comment (Cedev) [inline a10_neutron_lbaas/handler_base.py]: I'm surprised this isn't a context manager, like `with LATE_INIT_LOCK:`. I double checked, and it looks like the interface is acquire ... release.

**Comment by Cedev** [comment]
Looks good in the daylight. It'd be nice to test it with orchestration to see what it does.

File: a10_neutron_lbaas/a10_config.py (hunk 4 above)
Comment (dougwig) [inline a10_neutron_lbaas/a10_config.py]: Yeah, so, does a passed in config mean "use *this* config", or "use this object as if it was the file config.py" ?

File: a10_neutron_lbaas/handler_base.py (hunk 5 above)
Comment (dougwig) [inline a10_neutron_lbaas/handler_base.py]: Didn't know about that; cool, will switch.

File: a10_neutron_lbaas/a10_config.py (hunk 4 above)
Comment (mdurrant-b3) [inline a10_neutron_lbaas/a10_config.py]: A passed in config = "use this config".

**Comment by mdurrant-b3** [comment]
I haven't tested this by hand.

**Pull request merged**